device/transport/amqp/src/amqp.ts

```diff
@@ -74,54 +83,331 @@ export class Amqp extends EventEmitter implements Client.Transport, StableConnec
       // this will come with the retry logic.
       // this.emit('errorReceived', err);
     };
 
-    const c2dMessageListener = (msg) => {
+    this._c2dMessageListener = (msg) => {
       this.emit('message', msg);
     };
 
+    this._d2cErrorListener = (err) => {
+      debug('Error on the D2C link: ' + err.toString());
+      // this is the right thing to do (even better would be to emit 'error' directly but can't be done right now because the client does not have the proper error handling logic.)
+      // this will come with the retry logic.
+      // this.emit('errorReceived', err);
+    };
+
     /*Codes_SRS_NODE_DEVICE_AMQP_RECEIVER_16_008: [The `Amqp` object shall remove the listeners on `message` and `error` events of the underlying `ReceiverLink` when no-one is listening to its own `message` event.]*/
     this.on('removeListener', (eventName, eventCallback) => {
-      if (eventName === 'message') {
-        if (this._c2dLink) {
-          if (this.listeners('message').length === 0) {
-            this._c2dLink.detach((err) => {
-              if (err) {
-                debug('error detaching c2d link: ' + err.toString());
-              } else {
-                this._c2dLink.removeListener('error', c2dErrorListener);
-                this._c2dLink.removeListener('message', c2dMessageListener);
-                this._c2dLink = undefined;
-              }
-            });
+      if (eventName === 'message' && this.listeners('message').length === 0) {
+        debug('automatically detaching the link after the last message listener was removed');
+        this._fsm.handle('detachC2DLink', (err) => {
+          if (err) {
+            this.emit('errorReceived', err);
           }
-        } else {
-          debug('No active C2D link: nothing to do here.');
-        }
+        });
+      } else {
+        debug('removing listener for event: ' + eventName);
       }
     });
 
     /*Codes_SRS_NODE_DEVICE_AMQP_RECEIVER_16_003: [The `Amqp` object shall listen to the `message` and error events of the underlying `ReceiverLink` object when it has listeners on its `message` event.]*/
     this.on('newListener', (eventName, eventCallback) => {
       debug('registering new event handler for: ' + eventName);
       if (eventName === 'message') {
-        if (!this._c2dLink) {
-          this._amqp.attachReceiverLink(this._c2dEndpoint, null, (err, link) => {
-            this._c2dLink = link;
-            this._c2dLink.on('error', c2dErrorListener);
-            this._c2dLink.on('message', c2dMessageListener);
+        if (this.listeners('message').length === 0) {
+          debug('automatically attaching the link after the first message listener was registered');
+          this._fsm.handle('attachC2DLink', (err) => {
+            debug('C2D link attached automatically');
+            if (err) {
+              debug('error attaching the C2D link automatically: ' + err.toString());
+              this.emit('errorReceived', err);
```

Comment:
>this.emit('errorReceived', err); [](start = 14, length = 32)

thinking out loud, it would be nice if this could emit something that could be acted upon.  'errorReceived' is generic and the response is unknown.  'linkConnectionFailure' might be easier to act upon (disconnect then reconnect or something like that).

Reply:
2 things:
1/ i want to change this to the actual 'error' event. the trick is that any uncaught error is thrown so we have to make sure our error handling in the client is solid. this will happen later.
2/ this could be up for reconsideration (although it'd likely be a breaking change) but right now we're using the error type (and not the event name) to detect the type of failure. so the caller still knows what's going on, we're not throwing a generic thing that cannot be acted on.

---
In reply to: [141125512](https://github.com/Azure/azure-iot-sdk-node/pull/89#discussion_r141125512) [](ancestors = 141125512)